### PR TITLE
Fix `MultiMaterial` support in most primitives

### PR DESCRIPTION
**Pull request by ikerr**
This commit makes primitive geometries compatible with `MultiMaterial` by adding a single group to each primitive. Some primitives already supported `MultiMaterial` (e.g., `BoxGeometry` and `CylinderGeometry`).

I added a new example that can be used to test the new behavior.

### CODE REVIEW

File: src/geometries/CircleGeometry.js [hunk 1]
```diff
@@ -111,6 +111,9 @@ function CircleBufferGeometry( radius, segments, thetaStart, thetaLength ) {
 	this.addAttribute( 'normal', new Float32BufferAttribute( normals, 3 ) );
 	this.addAttribute( 'uv', new Float32BufferAttribute( uvs, 2 ) );
 
+	// add a group to the geometry. this will ensure multi material support
```
Comment (Mugen87) [inline src/geometries/CircleGeometry.js]: @ikerr Can you please insert after each comment an empty line?

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: src/geometries/CircleGeometry.js (hunk 1 above)
Comment (ikerr) [inline src/geometries/CircleGeometry.js]: done

**Comment by Mugen87** [comment]
This can be merged now 😉

**Comment by WestLangley** [comment]
Can you please explain:

In your example, all of the materials are the same in your materials array. How can you see if the code is working?

In what way are the geometries "not compatible" with `MultiMaterial` without this PR?

In what way are the geometries "compatible" with `MultiMaterial` with this PR?

**Comment by ikerr** [comment]
Hi @WestLangley ,

Without the PR, you'll get runtime errors when trying to render these primitives with a `MultiMaterial` assigned.

The materials in the array are the same; however, I think that only `BoxGeometry` makes use of the multiple materials (one for each face) - all of the other primitives only use the first material.

I could update the example to assign a different material to each face of the `BoxGeometry`, but the main intent of this PR was to ensure that assigning a `MultiMaterial` did not result in a runtime error.

**Comment by WestLangley** [comment]
I do not think the example is particularly useful as written.

I also do not think `groups` should be defined unnecessarily.

@Mugen87 Can this be fixed by modifying the renderer, instead?

**Comment by Mugen87** [comment]
The first problem is reproduced here: https://jsfiddle.net/oooao6qb/4/

You get a runtime error if you provide a `MultiMaterial` with less materials than expected. So a `BoxBufferGeometry` needs six materials but we assign just one.

I think this can be fixed if we change `.projectObject()`. https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1432-L1450

**Comment by Mugen87** [comment]
Um, i'm unable to reproduce a runtime error with a single `SphereGeometry`. https://jsfiddle.net/oooao6qb/5/

A `SphereBufferGeometry` don't provide any errors but nothing is rendered. I think this need some more investigation

https://jsfiddle.net/oooao6qb/6/

@WestLangley Thanks for intervening 😉

**Comment by ikerr** [comment]
I've looked into this some more, to figure out why regular `SphereGeometry` works while `SphereBufferGeometry` does not. The reason that `SphereBufferGeometry` does not work is because it has no groups, but has a `MultiMaterial` assigned, which trips up this bit of code in `WebGLRenderer` that expects `groups` to exist:

```
if ( material.isMultiMaterial ) {
  var groups = geometry.groups; // What if there are no groups? We'll get an error below:
  var materials = material.materials;

  for ( var i = 0, l = groups.length; i < l; i ++ ) {
    var group = groups[ i ];
    var groupMaterial = materials[ group.materialIndex ];

    if ( groupMaterial.visible === true ) {
      pushRenderItem( object, geometry, groupMaterial, _vector3.z, group );
    }
  }
}
```

*So why does regular `SphereGeometry` work?*

The short answer is that when `SphereGeometry` is converted to a `BufferGeometry` by the renderer, a single group is created, which allows the above code to work.

The longer answer is that:

1. `SphereGeometry` is actually created by passing a `SphereBufferGeometry` to `fromBufferGeometry()`
2. `fromBufferGeometry()` instances a bunch of `Face3` objects
3. In the case where the `BufferGeometry` does not contain any groups, `fromBufferGeometry()` invokes the `Face3` constructor with an undefined `materialIndex`
4. When the `Face3` constructor is called with `materialIndex = undefined`, it sets `this.materialIndex` to 0

So, `SphereGeometry` contains a bunch of `Face3` instances with `materialIndex = 0`. Now, when we render this:

1. The renderer converts the `SphereGeometry` to a `DirectGeometry` by calling `DirectGeometry.prototype.fromGeometry()`
2. `DirectGeometry.prototype.fromGeometry()` invokes `DirectGeometry.prototype.computeGroups()`, which creates groups based on material indices
3. If all of the material indices are undefined, no groups are created, but in our case, all of the material indices are 0, so a single group is created

Thus, `SphereGeometry` renders fine because a group is automatically created by the renderer.

We can likely modify the above renderer code to handle the case where a `MultiMaterial` is assigned but no groups are defined by doing something like this:

```
if ( material.isMultiMaterial && geometry.groups !== undefined ) {
  var groups = geometry.groups;
  var materials = material.materials;

  for ( var i = 0, l = groups.length; i < l; i ++ ) {
    var group = groups[ i ];
    var groupMaterial = materials[ group.materialIndex ];

    if ( groupMaterial.visible === true ) {
      pushRenderItem( object, geometry, groupMaterial, _vector3.z, group );
    }
  }
} else if ( material.isMultiMaterial && material.materials.length > 0 ) {
  // no groups are defined, so use the first material in the MultiMaterial
  pushRenderItem( object, geometry, materials[ 0 ], _vector3.z, null );
} else if ( !material.isMultiMaterial ) {
  pushRenderItem( object, geometry, material, _vector3.z, null );
}
```

or we can take the approach of this pull request and define a single group for each `*BufferGeometry` type.

**Comment by WestLangley** [comment]
>We can likely modify the above renderer code to handle the case where a MultiMaterial is assigned but no groups are defined

I suggest we instead throw a warning when a MultiMaterial is used but groups are not specified.

>Or we can take the approach of this pull request and define a single group for each *BufferGeometry type.

If we did that, when a users wants to specify a group, he has to first clear the default one, which should not be necessary in the first place.

**Comment by ikerr** [comment]
@WestLangley If we just throw a warning when a `MultiMaterial` is used but groups are not specified (vs. handling this case), and do not add a default group to the `*BufferGeometry` types, then none of the `*BufferGeometry` types will work "out of the box" with `MultiMaterial`. IMO that's not desirable.

**Comment by Mugen87** [comment]
I've update the PR #10698 with @WestLangley suggestions. Indeed, creating a default groups could be problematic if users want to create their own groups. Maybe it's better for the moment if we just provide warnings and avoid runtime errors like #10698 does now.

@ikerr Your right, the variance between `Geometry` and `BufferGeometry` is an inconsistency. Not sure, how to handle this in a straightforward way...

**Comment by WestLangley** [comment]
@ikerr What is your use case? It does not make sense to use MultiMaterial unless material indices (i.e.,  groups) are specified.

**Comment by ikerr** [comment]
In our engine we exclusively use `MultiMaterial` because it is the most general. Exclusively using `MultiMaterial` means that we don't need `material.isMultiMaterial` checks throughout our codebase. All of our "mesh" geometries have at least a single group defined, so `MultiMaterial` works fine for them. The problem we are having is that Three.js's primitives behave inconsistently with `MultiMaterial` - some work and some don't.

Taking a step back, I think this lack of consistency between the primitives and also between the "regular" and `BufferGeometry`-based versions is the real problem, as @Mugen87 noted.

I think the API would be cleaner if we could eliminate the following inconsistencies:

1. Some primitives, like `BoxBufferGeometry`, `BoxGeometry`, `CylinderBufferGeometry` and `CylinderGeometry`, have explicit groups defined while other primitives do not. It seems reasonable that either all of the primitives should define groups (or we interpret the lack of groups as a single "implicit" group), or none of them should.

2. The "regular", non-buffer geometries are created with an "implicit" group, because all of the material indices are initialized to 0 rather than undefined. That is, they render fine with a `MultiMaterial` applied whereas the `BufferGeometry` versions do not.

So I feel that there are two solutions: either we introduce explicit groups for each primitive (this PR in its current form), or we assume that a BufferGeometry has an "implicit" group when no explicit ones are present.

In a way, the renderer *already* assumes an "implicit" group for BufferGeometries that do not have any groups defined when it computes the draw range:

```
var rangeStart = geometry.drawRange.start * rangeFactor;                        
var rangeCount = geometry.drawRange.count * rangeFactor;                        
                                                                                
var groupStart = group !== null ? group.start * rangeFactor : 0;                
var groupCount = group !== null ? group.count * rangeFactor : Infinity;                                                                                         
var drawStart = Math.max( rangeStart, groupStart );                             
var drawEnd = Math.min( dataCount, rangeStart + rangeCount, groupStart + groupCount ) - 1;
                                                                                
var drawCount = Math.max( 0, drawEnd - drawStart + 1 );
```

It seems reasonable to extend this idea and make the renderer assume that the "implicit" group has a `materialIndex` set to 0, which it already does for non-buffer geometries.

**Comment by WestLangley** [comment]
>It seems reasonable to extend this idea and make the renderer assume that the "implicit" group has a materialIndex set to 0, which it already does for non-buffer geometries.

@ikerr Yes, that is the solution I was thinking of when I suggested we fix the renderer, instead.

Thing is, one could also argue that it does not make sense to use MultiMaterial unless material indices (i.e., groups) are specified. So a warning is appropriate.

I would be open to either.

**Comment by Benjamin-Dobell** [comment]
Can you guys please check-out #10093? Certainly wasn't my primary intention, but I may have already resolved this.

And I _really_ don't want to have to rebase again :wink:

**Comment by Benjamin-Dobell** [comment]
screen shot 2017-02-04 at 1 28 53 pm

This is what I'm seeing with my pull request. This is correct?

**Comment by mrdoob** [comment]
I agree with @WestLangley. Is there any way we can make the example a little bit more illustrative?

**Comment by ikerr** [comment]
Sorry for the tardiness on this, I've been sidetracked with other things and have been waiting to see how issue #10093 played out.

I'll modify this PR so that the renderer handles the existing primitives properly (vs. modifying the primitives), as we discussed above.

I'm not sure what to do about the example. We could just get rid of it, but it does give us a way to verify that `MultiMaterial` is compatible with the various primitives.

Does that sound good? @WestLangley @mrdoob @Mugen87

**Comment by WestLangley** [comment]
@ikerr Check and see to what extent your issue has been resolved in the `dev` branch.

**Comment by ikerr** [comment]
Looks like this was fixed in pull-request #10698

**Pull request closed without merging**